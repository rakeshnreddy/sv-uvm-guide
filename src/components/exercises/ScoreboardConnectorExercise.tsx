"use client";

import React, { useState, useRef, MouseEvent } from 'react';
import { Button } from '@/components/ui/Button';

interface Port {
  id: string;
  componentId: string;
  name: string; // e.g., "analysis_port", "data_imp"
  type: 'port' | 'export' | 'imp'; // Simplified type
  x: number; // Relative x within component
  y: number; // Relative y within component
  isOutput?: boolean; // true for port, false for imp/export (simplified)
}

interface Component {
  id: string;
  name: string;
  x: number;
  y: number;
  width: number;
  height: number;
  ports: Port[];
}

interface Connection {
  fromComponentId: string;
  fromPortId: string;
  toComponentId: string;
  toPortId: string;
}

const initialComponents: Component[] = [
  {
    id: 'monitor', name: 'UVM Monitor', x: 50, y: 150, width: 150, height: 80,
    ports: [{ id: 'mon_ap', componentId: 'monitor', name: 'trans_ap', type: 'port', x: 145, y: 40, isOutput: true }],
  },
  {
    id: 'scoreboard', name: 'Scoreboard', x: 350, y: 50, width: 150, height: 80,
    ports: [{ id: 'sb_imp_actual', componentId: 'scoreboard', name: 'actual_trans_imp', type: 'imp', x: 5, y: 40, isOutput: false }],
  },
  {
    id: 'coverage', name: 'Coverage Collector', x: 350, y: 250, width: 150, height: 80,
    ports: [{ id: 'cov_imp', componentId: 'coverage', name: 'observed_trans_imp', type: 'imp', x: 5, y: 40, isOutput: false }],
  },
];

const portRadius = 8;
const portStrokeColor = "#4A5568"; // gray-700
const portFillColor = "#E2E8F0"; // gray-200
const portFillColorHover = "#CBD5E0"; // gray-300

const ScoreboardConnectorExercise: React.FC = () => {
  const [components, /* setComponents */] = useState<Component[]>(initialComponents); // setComponents commented out
  const [connections, setConnections] = useState<Connection[]>([]);
  const [connectionResults, setConnectionResults] = useState<boolean[] | null>(null);
  const [score, setScore] = useState<number | null>(null);
  const [drawingLine, setDrawingLine] = useState<{ fromComponentId: string; fromPortId: string; x1: number; y1: number; x2: number; y2: number } | null>(null);
  const [selectedPort, setSelectedPort] = useState<{ componentId: string; portId: string } | null>(null);
  const svgRef = useRef<SVGSVGElement>(null);

  // Map of correct connections: output port -> valid input ports
  const referenceMap: Record<string, string[]> = {
    mon_ap: ['sb_imp_actual', 'cov_imp'],
  };

  const getPortAbsolutePosition = (componentId: string, portId: string) => {
    const component = components.find(c => c.id === componentId);
    const port = component?.ports.find(p => p.id === portId);
    if (component && port) {
      return { x: component.x + port.x, y: component.y + port.y };
    }
    return { x: 0, y: 0 };
  };

  const handlePortClick = (componentId: string, portId: string) => {
    const component = components.find(c => c.id === componentId);
    const port = component?.ports.find(p => p.id === portId);
    if (!component || !port) return;

    const { x, y } = getPortAbsolutePosition(componentId, portId);

    if (!selectedPort) {
      // Start drawing a line from this port
      setSelectedPort({ componentId, portId });
      setDrawingLine({ fromComponentId: componentId, fromPortId: portId, x1: x, y1: y, x2: x, y2: y });
    } else {
      // Finish drawing a line to this port, if it's not the same port
      if (selectedPort.componentId !== componentId || selectedPort.portId !== portId) {
        // Basic validation: prevent connecting output to output, input to input (simplified)
        const fromPortDetails = components.find(c => c.id === selectedPort.componentId)?.ports.find(p => p.id === selectedPort.portId);
        if (fromPortDetails && fromPortDetails.isOutput === port.isOutput) {
            alert("Cannot connect ports of the same type (e.g., output to output).");
            setSelectedPort(null);
            setDrawingLine(null);
            return;
        }

        // Prevent duplicate connections (simplified check)
        const newConnection: Connection = {
            fromComponentId: selectedPort.componentId,
            fromPortId: selectedPort.portId,
            toComponentId: componentId,
            toPortId: portId,
          };
        const alreadyExists = connections.some(conn =>
            (conn.fromPortId === newConnection.fromPortId && conn.toPortId === newConnection.toPortId) ||
            (conn.fromPortId === newConnection.toPortId && conn.toPortId === newConnection.fromPortId)
        );
        if (alreadyExists) {
            alert("This connection already exists.");
        } else {
             setConnections(prev => [...prev, newConnection]);
        }
      }
      setSelectedPort(null);
      setDrawingLine(null);
    }
  };

  const handleMouseMove = (event: MouseEvent<SVGSVGElement>) => {
    if (drawingLine && svgRef.current) {
      const svgRect = svgRef.current.getBoundingClientRect();
      setDrawingLine(prev => prev ? { ...prev, x2: event.clientX - svgRect.left, y2: event.clientY - svgRect.top } : null);
    }
  };

  const handleSvgClick = (event: MouseEvent<SVGSVGElement>) => {
    // If clicking on SVG canvas itself (not a port) and a line is being drawn, cancel it.
    if (selectedPort && event.target === svgRef.current) {
        setSelectedPort(null);
        setDrawingLine(null);
    }
  }

  const resetExercise = () => {
    setConnections([]);
    setConnectionResults(null);
    setScore(null);
  };

  const handleCheckConnections = () => {
    const results = connections.map(conn => {
      // Determine which port is the output to canonicalize connection direction
      const fromDetails = components
        .find(c => c.id === conn.fromComponentId)?.ports.find(p => p.id === conn.fromPortId);
      const toDetails = components
        .find(c => c.id === conn.toComponentId)?.ports.find(p => p.id === conn.toPortId);

      if (!fromDetails || !toDetails) return false;

      const outputPortId = fromDetails.isOutput ? conn.fromPortId : conn.toPortId;
      const inputPortId = fromDetails.isOutput ? conn.toPortId : conn.fromPortId;

      return referenceMap[outputPortId]?.includes(inputPortId) ?? false;
    });

    setConnectionResults(results);
    setScore(results.filter(Boolean).length);
  };

  return (
    <div className="w-full max-w-2xl mx-auto p-4 bg-white/10 backdrop-blur-lg border border-white/20 rounded-lg shadow-lg">
      <h3 className="text-lg font-semibold mb-4 text-center text-primary">Connect the UVM Components</h3>
      <p className="text-sm text-center text-muted-foreground mb-4">
        Click on an output port (e.g., an analysis_port on a monitor) and then click on a compatible input port
        (e.g., an analysis_imp on a scoreboard or coverage collector) to make a connection.
      </p>
      <svg
        ref={svgRef}
        width="100%"
        height="400"
        viewBox="0 0 600 400"
        className="border border-white/20 rounded bg-white/10"
        onMouseMove={handleMouseMove}
        onClick={handleSvgClick}
      >
        {/* Draw existing connections */}
        {connections.map((conn, index) => {
          const fromPos = getPortAbsolutePosition(conn.fromComponentId, conn.fromPortId);
          const toPos = getPortAbsolutePosition(conn.toComponentId, conn.toPortId);
          const strokeColor = connectionResults
            ? connectionResults[index]
              ? 'green'
              : 'red'
            : 'hsl(var(--accent))';
          return (
            <line
              key={`conn-${index}`}
              x1={fromPos.x} y1={fromPos.y}
              x2={toPos.x} y2={toPos.y}
              stroke={strokeColor}
              strokeWidth="2.5"
            />
          );
        })}

        {/* Draw line being currently drawn */}
        {drawingLine && (
          <line
            x1={drawingLine.x1} y1={drawingLine.y1}
            x2={drawingLine.x2} y2={drawingLine.y2}
            stroke="hsl(var(--foreground))"
            strokeWidth="2"
            strokeDasharray="4 4"
          />
        )}

        {/* Draw components and their ports */}
        {components.map(comp => (
          <g key={comp.id} transform={`translate(${comp.x}, ${comp.y})`}>
            <rect
              width={comp.width}
              height={comp.height}
              rx="5"
              fill="hsla(var(--primary-foreground), 0.1)"
              stroke="hsl(var(--primary))"
              strokeWidth="1.5"
            />
            <text x={comp.width / 2} y="20" textAnchor="middle" fontSize="12" fill="hsl(var(--foreground))">
              {comp.name}
            </text>
            {comp.ports.map(port => (
              <circle
                key={port.id}
                cx={port.x}
                cy={port.y}
                r={portRadius}
                fill={selectedPort?.portId === port.id ? "hsl(var(--accent))" : "hsl(var(--primary-foreground))"}
                stroke="hsl(var(--primary))"
                strokeWidth="1.5"
                className="cursor-pointer hover:fill-hsl(var(--accent))"
                onClick={(e) => { e.stopPropagation(); handlePortClick(comp.id, port.id);}}
                aria-label={`Port ${port.name} on ${comp.name}`}
              />
            ))}
          </g>
        ))}
      </svg>
       <div className="mt-4 flex justify-center gap-4">
        <Button
            onClick={resetExercise}
            variant="destructive"
        >
            Reset Connections
        </Button>
        <Button onClick={handleCheckConnections}>Check Connections</Button>
      </div>
      {score !== null && (
        <div className="mt-4 text-center space-y-2">
          <p>You got {score} out of {referenceMap['mon_ap'].length} connections correct.</p>
          <Button onClick={resetExercise} variant="secondary">Retry</Button>
        </div>
      )}
    </div>
  );
};

export default ScoreboardConnectorExercise;
