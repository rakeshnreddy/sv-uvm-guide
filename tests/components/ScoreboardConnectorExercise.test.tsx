import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, it, expect, vi } from 'vitest';
import ScoreboardConnectorExercise from '../../src/components/exercises/ScoreboardConnectorExercise';

describe('ScoreboardConnectorExercise', () => {
  it('scores correct connections', async () => {
    render(<ScoreboardConnectorExercise />);
    const monitorPort = screen.getByLabelText(/Port trans_ap on UVM Monitor/i);
    const scoreboardPort = screen.getByLabelText(/Port actual_trans_imp on Scoreboard/i);
    const coveragePort = screen.getByLabelText(/Port observed_trans_imp on Coverage Collector/i);

    await userEvent.click(monitorPort);
    await userEvent.click(scoreboardPort);
    await userEvent.click(monitorPort);
    await userEvent.click(coveragePort);

    await userEvent.click(screen.getByRole('button', { name: /check connections/i }));
    await screen.findByText(/You got 2 out of 2 connections correct/i);
  });

  it('warns on invalid connections and gives low score', async () => {
    const alertSpy = vi.spyOn(window, 'alert').mockImplementation(() => {});

    render(<ScoreboardConnectorExercise />);
    const scoreboardPort = screen.getByLabelText(/Port actual_trans_imp on Scoreboard/i);
    const coveragePort = screen.getByLabelText(/Port observed_trans_imp on Coverage Collector/i);

    await userEvent.click(scoreboardPort);
    await userEvent.click(coveragePort);
    expect(alertSpy).toHaveBeenCalled();

    await userEvent.click(screen.getByRole('button', { name: /check connections/i }));
    await screen.findByText(/You got 0 out of 2 connections correct/i);

    alertSpy.mockRestore();
  });
});
