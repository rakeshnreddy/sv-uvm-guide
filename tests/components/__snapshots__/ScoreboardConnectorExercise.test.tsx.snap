// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ScoreboardConnectorExercise component > renders ports with hover accent fill class 1`] = `
<DocumentFragment>
  <div
    class="w-full max-w-2xl mx-auto p-4 bg-white/10 backdrop-blur-lg border border-white/20 rounded-lg shadow-lg"
  >
    <h3
      class="text-lg font-semibold mb-4 text-center text-primary"
    >
      Connect the UVM Components
    </h3>
    <p
      class="text-sm text-center text-muted-foreground mb-4"
    >
      Click on an output port (e.g., an analysis_port on a monitor) and then click on a compatible input port (e.g., an analysis_imp on a scoreboard or coverage collector) to make a connection.
    </p>
    <svg
      class="border border-white/20 rounded bg-white/10"
      height="400"
      viewBox="0 0 600 400"
      width="100%"
    >
      <g
        transform="translate(50, 150)"
      >
        <rect
          fill="hsla(var(--primary-foreground), 0.1)"
          height="80"
          rx="5"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
          width="150"
        />
        <text
          fill="hsl(var(--foreground))"
          font-size="12"
          text-anchor="middle"
          x="75"
          y="20"
        >
          UVM Monitor
        </text>
        <circle
          aria-label="Port trans_ap on UVM Monitor"
          class="cursor-pointer hover:fill-[hsl(var(--accent))]"
          cx="145"
          cy="40"
          fill="hsl(var(--primary-foreground))"
          r="8"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
        />
      </g>
      <g
        transform="translate(350, 50)"
      >
        <rect
          fill="hsla(var(--primary-foreground), 0.1)"
          height="80"
          rx="5"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
          width="150"
        />
        <text
          fill="hsl(var(--foreground))"
          font-size="12"
          text-anchor="middle"
          x="75"
          y="20"
        >
          Scoreboard
        </text>
        <circle
          aria-label="Port actual_trans_imp on Scoreboard"
          class="cursor-pointer hover:fill-[hsl(var(--accent))]"
          cx="5"
          cy="40"
          fill="hsl(var(--primary-foreground))"
          r="8"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
        />
      </g>
      <g
        transform="translate(350, 250)"
      >
        <rect
          fill="hsla(var(--primary-foreground), 0.1)"
          height="80"
          rx="5"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
          width="150"
        />
        <text
          fill="hsl(var(--foreground))"
          font-size="12"
          text-anchor="middle"
          x="75"
          y="20"
        >
          Coverage Collector
        </text>
        <circle
          aria-label="Port observed_trans_imp on Coverage Collector"
          class="cursor-pointer hover:fill-[hsl(var(--accent))]"
          cx="5"
          cy="40"
          fill="hsl(var(--primary-foreground))"
          r="8"
          stroke="hsl(var(--primary))"
          stroke-width="1.5"
        />
      </g>
    </svg>
    <div
      class="mt-4 flex justify-center"
    >
      <button
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-destructive text-destructive-foreground hover:bg-destructive/90 h-10 px-4 py-2"
      >
        Reset Connections
      </button>
    </div>
  </div>
</DocumentFragment>
`;
